File: 0929-unique-email-addresses/0929-unique-email-addresses.java
class Solution {
    public int numUniqueEmails(String[] emails) {
        Set<String> uniqueEmails = new HashSet<>();
        for (String email : emails) {
            uniqueEmails.add(normalize(email));
        }
        return uniqueEmails.size();
    }
    
    private String normalize(String email) {
        String[] args = email.split("@");
        String localName = args[0];
        int plusIdx = localName.indexOf("+");
        if (plusIdx != -1) {
            localName = localName.substring(0, plusIdx);
        }
        localName = localName.replaceAll("\\.", "");
        String domainName = args[1];
        return localName + "@" + domainName;
    }
}

File: 0929-unique-email-addresses/0929-unique-email-addresses.kt
class Solution {
    fun numUniqueEmails(emails: Array<String>): Int {
        val unique = HashSet<String>()
        for (email in emails) {
            unique.add(normalize(email))
        }
        return unique.size
    }

    private fun normalize(email: String) : String {
        var (local, domain) = email.split("@")
        val plus = local.indexOf("+")
        if (plus != -1) {
            local = local.substring(0, plus)
        }
        local = local.replace(".", "")
        return "$local@$domain"
    }
}